File: ios/AtlasKit/Sources/AtlasKit/Domain/Domains.swift
// The domain axis: what settles a claim in this corner of the world. The port
// of `lib/curriculum/domains.ts`.
//
// `NodeKind` says what the learner must be able to *do* — tell instances apart,
// run a sequence, predict what changes. It deliberately does not say what
// subject the node belongs to, and it structurally cannot: "settle it by
// deriving it" and "settle it by reading the source in context" are not two
// kinds of doing, they are two kinds of *warrant*. That is this axis.
//
// The two are orthogonal almost everywhere — a `procedure` exists in every
// domain — which is why this is a second field rather than more `NodeKind`
// cases. `general` behaves exactly as the pre-domain engine did.

/// What settles a claim here.
///
/// The discriminator is the *warrant*, never the subject: history and religion
/// are both `interpretive` because their epistemics are identical — a text, a
/// context, a contested reading, provenance that matters.
///
/// `performative` and `craft` split on one question: **can the app observe the
/// production?** It can hear a spoken sentence and it cannot see a sofa. That
/// decides whether the app grades the work or debriefs the learner about work
/// it never saw, which is a different ladder and a different gate.
public enum Domain: String, Codable, Sendable, CaseIterable {
    /// Derivation from stated rules. Mathematics, logic, theory.
    case formal
    /// It runs, or it fails. Code, engineering, protocols.
    case executable
    /// Measurement, carrying uncertainty. The sciences, medicine.
    case empirical
    /// A source, read in context, contested. History, religion, law.
    case interpretive
    /// Production under real conditions, which the app can observe. Language.
    case performative
    /// Production the app cannot observe — a physical artifact. Woodwork.
    case craft
    /// No stance. The pre-domain default, and the fallback for anything unclear.
    case general
}

/// A domain off the wire, or `general` for anything this build does not know.
/// Lenient for the same reason `asNodeKind` is: a map built by a newer server
/// must still draw here.
public func asDomain(_ raw: String?) -> Domain {
    Domain(rawValue: raw ?? "") ?? .general
}

/// How a domain changes the ladder.
///
/// Two rule shapes, because two genuinely different things happen. `add` merges
/// rungs into whatever the node's kind already runs — this is what gives a
/// `concept` node in a `formal` map something to actually compute, since
/// `phasePlans[.concept]` carries no execution rung at all. `plan` replaces the
/// kind's ladder outright, for the domains where kind stops being the thing
/// that decides: explaining the preterite in your own words is not speaking
/// Spanish, so `performative` takes Socratic, Feynman and Crucible *off*.
public enum DomainPlanRule: Sendable {
    case add([Phase])
    case plan([Phase])
}

public let domainPlans: [Domain: DomainPlanRule] = [
    // A claim is settled by deriving it, so the learner has to derive one.
    .formal: .add([.trace, .perform, .drill]),
    // Same three: reading about a program is not running one.
    .executable: .add([.trace, .perform, .drill]),
    // Settled by measurement, so the test is whether it forecasts one.
    .empirical: .add([.predict, .perform]),
    // A claim is settled by a source read in context, so the learner reads one
    // and then carries the disagreement honestly.
    .interpretive: .add([.provenance, .steelman]),
    // Production the app can hear. Nothing here is served by explaining the
    // language in prose.
    .performative: .plan([.consume, .discriminate, .drill, .produce, .recall, .retain]),
    // Work the app never sees. Predict runs the failures in simulation before
    // the learner runs them in oak; Perform is the debrief of work already done.
    .craft: .plan([.consume, .discriminate, .predict, .perform, .retain]),
]

/// The ladder a `(kind, domain)` pair runs, before anything is stored.
///
/// The merge is a *filter over `Phase.allCases`* rather than a concatenation,
/// which is what makes the subsequence invariant hold by construction: whatever
/// the two tables ask for comes back in canonical order with no duplicates.
///
/// Then the two cost axes only ever *remove* rungs: a support concept drops the
/// depth phases, and Socratic is dropped for an easy concept (and for a support
/// one unless it is hard). Defaults reproduce the pre-axes ladder. Mirrors
/// `resolvePlan` in `phases.ts`.
public func resolvePlan(
    _ kind: NodeKind, _ domain: Domain,
    _ importance: NodeImportance = .core, _ difficulty: NodeDifficulty = .medium
) -> [Phase] {
    let base = phasePlans[kind] ?? legacyPhasePlan
    var want: Set<Phase>
    switch domainPlans[domain] {
    case nil: want = Set(base)
    case .plan(let replacement): want = Set(replacement)
    case .add(let extra): want = Set(base).union(extra)
    }
    let support = importance == .support
    if support { want.subtract([.feynman, .connect, .crucible, .drill, .steelman]) }
    if difficulty == .easy || (support && difficulty != .hard) { want.remove(.socratic) }
    return Phase.allCases.filter { want.contains($0) }
}

/// What settles a claim about this topic, read off the map it produced.
///
/// The port of `topicDomainOf` in `lib/curriculum/domains.ts`. There is no
/// topic-level domain column to read: it was dropped precisely because it could
/// only drift from the nodes it claimed to summarise. The commonest non-general
/// domain among the nodes IS the topic's domain, and a map with none is
/// `general` — which is what a mixed map (an ML course with `formal` and
/// `executable` nodes) should fall back to for a question asked about the whole
/// topic.
public func topicDomainOf(_ nodes: [ConceptNode]) -> Domain {
    var counts: [Domain: Int] = [:]
    for node in nodes {
        guard let domain = node.domain, domain != .general else { continue }
        counts[domain, default: 0] += 1
    }
    // Walked in `allCases` order, first past the post — so a tie always breaks
    // the same way and the same map always answers the same thing. A dictionary
    // walk would not, and the answer reaches a cache key.
    var best: Domain = .general
    var most = 0
    for domain in Domain.allCases where (counts[domain] ?? 0) > most {
        best = domain
        most = counts[domain] ?? 0
    }
    return best
}

File: lib/curriculum/domains.ts
// The domain axis: what counts as evidence in this corner of the world.
//
// `NodeKind` says what the learner must be able to *do* — tell instances apart,
// run a sequence, predict what changes. It deliberately does not say what
// subject the node belongs to (`KIND_RULE` in `lib/server/generate/map.ts`), and
// it structurally cannot: "settle it by deriving it" and "settle it by reading
// the source in context" are not two kinds of doing, they are two kinds of
// *warrant*. That is this axis.
//
// The two are orthogonal almost everywhere — a `procedure` exists in every
// domain — which is why this is a second field rather than more `NodeKind`
// values. `general` returns nothing everywhere, exactly as `concept` does for
// kinds, so a map built before domains existed behaves byte-identically and its
// cached rows stay addressable (see `domainOf` in `lib/server/job.ts`).
//
// Type-only import: `PhaseId` is erased at compile time, so `phases.ts` can
// import DOMAIN_PLAN from here without a runtime cycle.
import type { PhaseId } from "./phases";

/**
 * What settles a claim here.
 *
 * The discriminator is the *warrant*, never the subject: history and religion
 * are both `interpretive` because their epistemics are identical — a text, a
 * context, a contested reading, provenance that matters. The topic string
 * already carries the subject into every prompt; this carries the stance.
 *
 * `performative` and `craft` split on one question: **can the app observe the
 * production?** It can hear a spoken Spanish sentence and it cannot see a sofa.
 * That is not a shade of the same thing — it decides whether the app grades the
 * work or debriefs the learner about work it never saw, which is a different
 * ladder and a different gate.
 */
export type Domain =
  /** Derivation from stated rules. Mathematics, logic, theory. */
  | "formal"
  /** It runs, or it fails. Code, engineering, protocols. */
  | "executable"
  /** Measurement, carrying uncertainty. The sciences, medicine, econometrics. */
  | "empirical"
  /** A source, read in context, contested. History, religion, law, literature. */
  | "interpretive"
  /** Production under real conditions, which the app can observe. Language. */
  | "performative"
  /** Production the app cannot observe — a physical artifact. Woodwork, cooking. */
  | "craft"
  /** No stance. The pre-domain default, and the fallback for anything unclear. */
  | "general";

export const DOMAINS = [
  "formal",
  "executable",
  "empirical",
  "interpretive",
  "performative",
  "craft",
  "general",
] as const;

/** Lenient in both directions, like `asNodeKind`: anything unrecognized reads as
 *  `general`, so an older map, a newer server, or a model that invented a value
 *  degrades to the pre-domain behaviour rather than to nonsense. */
export function asDomain(raw: unknown): Domain {
  return (DOMAINS as readonly string[]).includes(raw as string)
    ? (raw as Domain)
    : "general";
}

/**
 * How a domain changes the ladder — the second of the four layers a domain
 * reaches (map semantics, ladder, prompt stance, verification).
 *
 * Two rule shapes, because two genuinely different things happen:
 *
 * - `add` merges rungs into whatever the node's kind already runs. This is the
 *   fix for the structural bug: `PHASE_PLAN.concept` contains no execution rung
 *   at all, so a `concept` node in a `formal` map could reach `mastered` without
 *   the learner ever computing anything.
 * - `plan` replaces the kind's ladder outright, for the domains where kind
 *   stops being the thing that decides. Explaining the preterite in your own
 *   words is not speaking Spanish: `performative` takes Socratic, Feynman and
 *   Crucible *off*, and no amount of prompt steering makes those rungs earn
 *   their place on a language map.
 *
 * Removal is safe against all three invariants in `tests/curriculum.test.ts`:
 * a shorter list is still a subsequence of `PHASE_ORDER`, and every rule here
 * still opens on Consume and closes on Retain.
 */
export type DomainPlanRule =
  { readonly add: readonly PhaseId[] } | { readonly plan: readonly PhaseId[] };

export const DOMAIN_PLAN: Partial<Record<Domain, DomainPlanRule>> = {
  // A claim is settled by deriving it, so the learner has to derive one. Watch
  // the derivation run, run it on a real case, then run it fast.
  formal: { add: ["trace", "perform", "drill"] },
  // Same three, for the same reason: reading about a program is not running one.
  executable: { add: ["trace", "perform", "drill"] },
  // Settled by measurement, so the test is whether it forecasts the measurement
  // before it is taken, and whether the learner can actually take it.
  empirical: { add: ["predict", "perform"] },
  // A claim is settled by a source read in context, so the learner reads one
  // (Provenance) and then carries the disagreement honestly (Steelman).
  interpretive: { add: ["provenance", "steelman"] },
  // Production the app can hear. Discriminate carries the boundaries (ser/estar,
  // preterite/imperfect), Drill carries automaticity, Produce is the live
  // utterance, Retain carries the deck — and nothing here is served by
  // explaining the language in prose.
  performative: {
    plan: ["consume", "discriminate", "drill", "produce", "recall", "retain"],
  },
  // Work the app never sees. Predict runs the failures in simulation before the
  // learner runs them in oak; Perform is the debrief of work already done.
  craft: { plan: ["consume", "discriminate", "predict", "perform", "retain"] },
};
